sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -931,6 +931,14 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression {
         $evPrim = $result.copy();
       """
   }
+
+  override def sql: Option[String] = {
+    if (foldable) {
+      Literal.create(eval(), dataType).sql
+    } else {
+      child.sql.map(childSQL => s"CAST($childSQL AS ${dataType.sql})")
```

Comment:
How about we do not evaluate the expr if it is foldable? I think it does not really give us any benefit if we evaluate it at here.